indico/modules/events/papers/controllers/display.py

```diff
@@ -17,39 +17,145 @@
-from indico.modules.events.papers.operations import create_paper_revision
-from indico.web.util import jsonify_form, jsonify_data
+from indico.modules.events.papers.controllers.base import RHPaperBase
+from indico.modules.events.papers.forms import PaperSubmissionForm, PaperCommentForm, build_review_form
+from indico.modules.events.papers.models.comments import PaperReviewComment
+from indico.modules.events.papers.models.files import PaperFile
+from indico.modules.events.papers.models.reviews import PaperReviewType, PaperReview, PaperTypeProxy
+from indico.modules.events.papers.operations import (create_paper_revision, create_review, create_comment,
+                                                     update_comment, delete_comment, update_review)
+from indico.modules.events.papers.views import WPDisplayPapersBase, render_paper_page
+from indico.web.flask.templating import get_template_module
+from indico.web.util import jsonify_form, jsonify_data, jsonify
 
 
-class RHSubmitPaper(RHPapersBase):
+class RHSubmitPaper(RHPaperBase):
+    def _process(self):
+        form = PaperSubmissionForm()
+        if form.validate_on_submit():
+            create_paper_revision(self.contribution, session.user, form.files.data)
+            return jsonify_data(flash=False)
+        return jsonify_form(form, form_header_kwargs={'action': request.relative_url})
+
+
+class RHPaperTimeline(RHPaperBase):
+    def _process(self):
+        comment_form = PaperCommentForm(paper=self.paper, user=session.user, formdata=None)
+        return WPDisplayPapersBase.render_template('paper.html', self._conf, paper=self.paper,
+                                                   comment_form=comment_form, review_form=None)
+
+
+class RHDownloadPaperFile(RHPaperBase):
+    """Download a paper file"""
+
     normalize_url_spec = {
         'locators': {
-            lambda self: self.contribution
+            lambda self: self.file
         }
     }
 
     def _checkParams(self, params):
-        RHPapersBase._checkParams(self, params)
-        self.contribution = Contribution.get_one(request.view_args['contrib_id'], is_deleted=False)
+        RHPaperBase._checkParams(self, params)
+        self.file = PaperFile.get_one(request.view_args['file_id'])
+
+    def _process(self):
+        return self.file.send()
+
+
+class RHSubmitPaperReview(RHPaperBase):
+    """Review an paper in a specific reviewing type"""
+
+    normalize_url_spec = {
+        'locators': {
+            lambda self: self.paper,
+            lambda self: self.type_
+        }
+    }
+
+    def _check_paper_protection(self):
+        if self.paper.last_revision.get_reviews(user=session.user, group=self.type_.instance):
+            return False
+        return self.paper.can_review(session.user, check_state=True)
+
+    def _checkParams(self, params):
+        RHPaperBase._checkParams(self, params)
+        self.type_ = PaperTypeProxy(PaperReviewType(request.view_args['type_id']))
```

Comment:
No need for the underscore since `self.type` wouldn't shadow `type()` - and I'd go for `review_type` here too (same for other places which I'm too lazy to look for now ;))